### PR TITLE
✨ 360 Feature discovery

### LINKED ISSUE #29591: [Story 360] Feature discovery UI
Implement the "move your device" feature discovery UI.

image (1)

cc @ampproject/wg-stories

### PR DESCRIPTION
**Pull request by processprocess**
Context / Fixes #29591
[Demo](https://discovery-ec239.web.app/examples/amp-story-360.amp.html)

Sep-24-2020 13-12-23

cc @hongcatlover

### CODE REVIEW

File: extensions/amp-story-360/0.1/amp-story-360.js [hunk 1]
```diff
@@ -385,33 +381,39 @@ export class AmpStory360 extends AMP.BaseElement {
...
+      this.win.addEventListener('deviceorientation', (e) => {
+        if (this.isReady_ && this.isOnActivePage_) {
+          rafTimeout && this.win.cancelAnimationFrame(rafTimeout);
+          rafTimeout = this.win.requestAnimationFrame(() =>
+            this.onDeviceOrientation_(e)
+          );
+        }
+      });
+      // Display discovery animation.
+      if (this.isOnActivePage_) {
```
Comment (gmajoulet) [inline extensions/amp-story-360/0.1/amp-story-360.js:406]: About when this animation is triggered:
- Do we want to trigger it on each 360 media of a Story?
- Do we re-trigger it when restarting the page (navigation forward then backward)?
- If I understand correctly the animation plays on `layoutCallback` or upon gyroscope permission grant. Can we change it so it's triggered when the page becomes active?

File: extensions/amp-story-360/0.1/amp-story-360.js [hunk 2]
```diff
@@ -385,33 +381,39 @@ export class AmpStory360 extends AMP.BaseElement {
...
+        const discoveryHTML = htmlFor(this.element)`
```
Comment (gmajoulet) [inline extensions/amp-story-360/0.1/amp-story-360.js:407]: Nit: plz declare this template at the top of this file

File: extensions/amp-story-360/0.1/amp-story-360.js (hunk 1 above)
Comment (t0mg) [reply extensions/amp-story-360/0.1/amp-story-360.js:406]: +1 and also: can we have a way to disable it ? the stories I'm working on will have text right where the animation runs, it'll likely not look good. 

Besides, having an attribute to  toggle that animation might also allow creators to only show it on eg the first page that has a 360 player and turn it off on subsequent pages. Because basing that only on permission grant would probably not work as expected on androd.

Comment (gmajoulet) [reply extensions/amp-story-360/0.1/amp-story-360.js:406]: We can file a new ticket for a way to opt-out, I'd like PM/UX to contribute to the conversation.

Comment (newmuis) [reply extensions/amp-story-360/0.1/amp-story-360.js:406]: This really feels like part of the story education system, would be interesting to align requirements with that

Comment (t0mg) [reply extensions/amp-story-360/0.1/amp-story-360.js:406]: An option to swap the position of the animation would work as well. Or have it vertically centered with a darker overlay that would dim out the text layer in the content no matter its location.

Comment (processprocess) [reply extensions/amp-story-360/0.1/amp-story-360.js:423]: Thanks for this great feedback :) 

> Do we want to trigger it on each 360 media of a Story?
> Do we re-trigger it when restarting the page (navigation forward then backward)?

If we assume the user will recognize the feature once they are aware of it, we can show it on the first active 360 page while the feature is enabled.

> Can we have a way to disable it?

If we consider this optional we shouldn’t have it at all. 
Anything in between creates an inconsistent UI experience.

> This really feels like part of the story education system, would be interesting to align requirements with that.

This is component level rather than story level so I think it’s okay for it to be visually tailored to the component. 
I agree that it would be good to be sure this is aligned with our education experiences

> An option to swap the position of the animation would work as well. Or have it vertically centered with a darker overlay that would dim out the text layer in the content no matter its location.

Definitely  needs  contrast against content behind it. The layout and look of it needs to be consistent. 
It’s important to see some of the background clearly so the user sees the content moving.

cc @hongcatlover

Comment (newmuis) [reply extensions/amp-story-360/0.1/amp-story-360.js:426]: SGTM!

For the record, I wasn't referring to the UX re: aligning with story education, but more how we keep track of whether or not to show this to users.  We should hook into the education system's capabilities around (a) determining whether the user has seen this before and/or should see it (again), and (b) allowing different players to have different configurations for when/how it should be shown.

File: extensions/amp-story-360/0.1/amp-story-360.js [hunk 3]
```diff
@@ -385,33 +398,36 @@ export class AmpStory360 extends AMP.BaseElement {
...
+        if (this.isReady_ && this.isOnActivePage_) {
+          rafTimeout && this.win.cancelAnimationFrame(rafTimeout);
+          rafTimeout = this.win.requestAnimationFrame(() =>
+            this.onDeviceOrientation_(e)
+          );
+        }
+      });
+      // Display discovery animation.
+      if (this.isOnActivePage_) {
+        const storyEl = this.win.document.querySelector('amp-story');
```
Comment (gmajoulet) [inline extensions/amp-story-360/0.1/amp-story-360.js:427]: Why did you move away from having it as part of the component? I think it's weird to have it at a random position in the DOM, especially as we can create many of them and never do clean up.

File: extensions/amp-story-360/0.1/amp-story-360.js (hunk 1 above)
Comment (gmajoulet) [reply extensions/amp-story-360/0.1/amp-story-360.js:426]: That SGTM as well, please ping us when you get a chance to write the code!

File: extensions/amp-story-360/0.1/amp-story-360.js (hunk 3 above)
Comment (processprocess) [reply extensions/amp-story-360/0.1/amp-story-360.js:427]: It's the same problem as the activate button #30420. 
Content overlaps it since it's a child of `amp-story-grid-layer`.

I converted this to draft since this needs work.
